chasm/registrable_component.go

```diff
@@ -67,24 +68,68 @@ func WithShardingFn(
+// This option must be specified if the archetype uses the Visibility component.
+func WithBusinessIDAlias(
+	alias string,
+) RegistrableComponentOption {
+	return func(rc *RegistrableComponent) {
+		if rc.searchAttributesMapper == nil {
+			rc.searchAttributesMapper = &VisibilitySearchAttributesMapper{
+				aliasToField:       make(map[string]string),
+				fieldToAlias:       make(map[string]string),
+				saTypeMap:          make(map[string]enumspb.IndexedValueType),
+				systemAliasToField: make(map[string]string),
+			}
+		}
+		if rc.searchAttributesMapper.systemAliasToField == nil {
+			rc.searchAttributesMapper.systemAliasToField = make(map[string]string)
+		}
+		if _, ok := rc.searchAttributesMapper.aliasToField[alias]; ok {
+			//nolint:forbidigo
+			panic(fmt.Sprintf("registrable component validation error: business ID alias %q is already defined as a search attribute", alias))
+		}
+		if _, ok := rc.searchAttributesMapper.systemAliasToField[alias]; ok {
+			//nolint:forbidigo
+			panic(fmt.Sprintf("registrable component validation error: business ID alias %q is already defined as a system search attribute", alias))
+		}
+		rc.searchAttributesMapper.systemAliasToField[alias] = sadefs.WorkflowID
+		rc.searchAttributesMapper.fieldToAlias[sadefs.WorkflowID] = alias
+		rc.searchAttributesMapper.saTypeMap[sadefs.WorkflowID] = enumspb.INDEXED_VALUE_TYPE_KEYWORD
+	}
+}
+
 func WithSearchAttributes(
 	searchAttributes ...SearchAttribute,
 ) RegistrableComponentOption {
 	return func(rc *RegistrableComponent) {
 		if len(searchAttributes) == 0 {
 			return
 		}
-		rc.searchAttributesMapper = &VisibilitySearchAttributesMapper{
-			aliasToField: make(map[string]string, len(searchAttributes)),
-			fieldToAlias: make(map[string]string, len(searchAttributes)),
-			saTypeMap:    make(map[string]enumspb.IndexedValueType, len(searchAttributes)),
+
+		if rc.searchAttributesMapper == nil {
+			rc.searchAttributesMapper = &VisibilitySearchAttributesMapper{
+				aliasToField:       make(map[string]string, len(searchAttributes)),
+				fieldToAlias:       make(map[string]string, len(searchAttributes)),
+				saTypeMap:          make(map[string]enumspb.IndexedValueType, len(searchAttributes)),
+				systemAliasToField: make(map[string]string),
+			}
 		}
 
 		for _, sa := range searchAttributes {
 			alias := sa.definition().alias
 			field := sa.definition().field
 			valueType := sa.definition().valueType
 
+			if sadefs.IsSystem(alias) || sadefs.IsReserved(alias) {
+				//nolint:forbidigo
+				panic(fmt.Sprintf("registrable component validation error: CHASM search attribute alias %q is a system or reserved search attribute", alias))
+			}
+
+			if _, ok := rc.searchAttributesMapper.systemAliasToField[alias]; ok {
+				//nolint:forbidigo
+				panic(fmt.Sprintf("registrable component validation error: CHASM search attribute alias %q is already defined as a system search attribute alias", alias))
```

Comment:
The error messages below says "search attribute alias", not "CHASM search attribute alias". Be consistent.